Readme.md
#Mongoose-TTL Plugin

Provides time-to-live support for [Mongoose](http://mongoosejs.com).

[![Build Status](https://secure.travis-ci.org/aheckmann/mongoose-ttl.png)](http://travis-ci.org/aheckmann/mongoose-ttl)

Options:

  - ttl: the time each doc should live in the db (default 60 seconds)
  - interval: how often the expired doc reaper runs (default 5 mins)
  - reap: enable the expired doc reaper (default true)
  - onReap: callback passed to reaper execution

Example:

```js
var schema = new Schema({..});
schema.plugin(ttl, { ttl: 5000 });
```

The ttl option supports the ms module by [guille](https://github.com/guille) meaning
we can specify ttls with friendlier syntax. Example:

```
 value     milliseconds
========================
 '2d'      172800000
 '1.5h'    5400000
 '1h'      3600000
 '1m'      60000
 '5s'      5000
 '500ms'   500
 100       100
```

The expired document reaper can be disabled by passing `reap: false`.
Useful when working in multi-core environments when we only want one
process executing it.

```js
var schema = new Schema({..});
schema.plugin(ttl, { ttl: 5000, reap: false });
var Cache = db.model('Cache', schema);
if (isMyWorker) Cache.startTTLReaper();
```

The reaper can also be stopped.

```js
Cache.stopTTLReaper();
```

Time-to-live is specified at the collection level, however
it can also be overridden for a given document.

```js
var cache = new Cache;
cache.ttl = '2m' // lives for two minutes
cache.save();
```

We can also reset the ttl for a given document to its
default plugin state.

```js
cache.resetTTL();
```
## Mongoose Version
>= 2.5.2

### MongoDB TTL collections
MongoDB is getting official support for TTL collections [soon-ish](https://github.com/mongodb/mongo/commit/25bdc679a0e559d64ec7f22b0468cf5b1671c4e7) at which point this plugin will need to change.

[LICENSE](https://github.com/aheckmann/mongoose-ttl/blob/master/LICENSE)


History.md

0.0.3 / 06-25-2012
==================

  * added; compatibility with node 0.8 and mongoose 3.x

0.0.2 / 02-09-2012
==================

  * changed; use schema!init event (requires mongoose >= 2.5.2)

0.0.1 / 01-24-2012
==================

  * initial release


test/index.js

var mongoose = require('mongoose')
  , Schema = mongoose.Schema
  , ttl = require('../')
  , assert = require('assert')

mongoose.connect('localhost', 'mongoose_test_ttl');

var ttl_ = 1000
  , interval = 1500
  , start = Date.now()

Date.prototype.inspect = function () {
  return '' + this.getTime();
}

var reaperCb = false;
var reaper = new (require('events').EventEmitter);
function log (err) {
  reaperCb = true;
  reaper.emit('reap');
}

var ThingSchema = new Schema({ name: String });
ThingSchema.plugin(ttl, { ttl: ttl_, interval: interval, onReap: log });
var T = mongoose.model('Thing', ThingSchema);

var indexCreated = false;
T.on('index', function () {
  indexCreated = true;
});

describe('ttl', function () {
  before(function (done) {
    mongoose.connection.on('open', function () {
      mongoose.connection.db.dropDatabase(function (err) {
        if (err) return done(err);
        assert.ok(T.__ttl);
        done();
      });
    });
  });

  describe('Model.find', function () {

    it('includes active ttls', function (done) {
      var t = new T({ name: 'vanilla latte' });
      t.save(function (err) {
        if (err) return done(err);

        T.find({ name: /latte$/ }).exec(function (err, docs) {
          if (err) return done(err);
          assert.equal(1, docs.length);
          assert.equal(t.id,docs[0].id);
          done();
        });
      });
    });

    it('excludes expired ttls', function (done) {
      var t = new T({ name: 'Tha Blob' });
      t.save(function (err) {
        if (err) return done(err);

        setTimeout(function () {
          T.find({ name: 'Tha Blob' }).exec(function (err, docs) {
            if (err) return done(err);
            assert.equal(0, docs.length);
            done();
          });
        }, ttl_ );
      });
    });

  });

  describe('Model queries', function () {
    var t;
    before(function (done) {
      T.create({ name: 'Tha Blob' }, function (err, t_) {
        assert.strictEqual(null, err);
        t = t_;
        done();
      });
    });

    after(function (done) {
      t.remove(done);
    });

    it('should not break existing query methods', function (done) {
      var pending = 7;

      T.where('name').exists().exec(function (err, docs) {
        assert.strictEqual(null, err);
        assert.equal(docs.length,1);
        assert.equal(t.id,docs[0].id);
        if (!--pending) done();
      });

      T.findOne({ name: 'Tha Blob' }, function (err, doc) {
        assert.strictEqual(null, err);
        assert.equal(t.id,doc.id);
        if (!--pending) done();
      });

      T.count({}, function (err, n) {
        assert.strictEqual(null, err);
        assert.equal(n,1);
        if (!--pending) done();
      });

      T.find().where('name', 'Tha Blob').exec(function (err, docs) {
        assert.strictEqual(null, err);
        assert.equal(docs.length,1);
        assert.equal(t.id,docs[0].id);
        if (!--pending) done();
      });

      T.distinct('name', { _id: { $exists: true }}, function (err, t) {
        assert.strictEqual(null, err);
        assert.equal(t.length,1);
        assert.equal(t[0],'Tha Blob');
        if (!--pending) done();
      });

      T.$where('this.name == "Tha Blob"').exec(function (err, docs) {
        assert.strictEqual(null, err);
        assert.equal(1, docs.length);
        assert.equal(t.id,docs[0].id);
        if (!--pending) done();
      });

      T.find({ __ttl: { $exists: true } }).exec(function (err, docs) {
        assert.strictEqual(null, err);
        assert.equal(docs.length,1);
        assert.equal(t.id,docs[0].id);
        if (!--pending) done();
      });
    });

    it('should convert __ttl params to $and arrays', function () {
      var query = T.find({ __ttl: { $exists: true } });
      var c = query._conditions;
      assert.ok(c.hasOwnProperty('$and'));
      assert.equal(c.$and.length,2);
      c.$and.forEach(function (arg) {
        assert.ok(arg.hasOwnProperty('__ttl'));
      });
    });
  });

  describe('virtuals', function () {
    describe('getters', function () {
      it('should return the current ttl', function () {
        var t = new T;
        assert.ok(t.ttl instanceof Date);
      });
    });

    describe('setters', function () {
      it('should set ttl to the value + the current date', function () {
        var now = Date.now();
        var t = new T;

        assert.ok((t.ttl - (now + 1000)) < 4);
        t.ttl = '2s';
        assert.ok((t.ttl - (now + 2000)) < 4);
        t.ttl = 3001;
        assert.ok((t.ttl - (now + 3001)) < 4);
        t.ttl = '500ms';
        assert.ok((t.ttl - (now + 500)) < 4);
      });

      it('should honor the ttl', function (done) {
        this.timeout(5000);
        var t = new T;
        t.ttl = 2000;
        var pending = 2;

        t.save(function (err) {
          assert.strictEqual(null, err);
        });

        setTimeout(function () {
          T.findById(t, function (err, b) {
            assert.strictEqual(null, err);
            assert.ok(b);
            finish();
          });
        }, 1200);

        setTimeout(function () {
          T.findById(t, function (err, b) {
            assert.strictEqual(null, err);
            assert.ok(!b);
            finish();
          });
        }, 2200);

        function finish () {
          if (--pending) return;
          done();
        }
      });
    });

  });

  describe('resetTTL', function () {
    it('should restore the plugin defaults', function () {
      var now = Date.now();
      var t = new T;
      t.ttl = '2d';
      assert.ok((t.ttl - (now + 172800000)) < 4);
      t.resetTTL();
      assert.ok((t.ttl - (now + 1000)) < 4);
    });
  });

  describe('reaper', function () {

    it('should remove all expired docs', function (done) {
      this.timeout(interval * 3);
      T.create({ name: 'Interval reaped' }, function (err, t) {
        assert.strictEqual(null, err);

        setTimeout(function () {
          reaper.once('reap', function () {
            T.collection.count(function (err, count) {
              assert.strictEqual(err, null);
              assert.equal(0, count);
              done();
            });
          });
        }, interval + 10);

      });
    });

    it('should fire the optional callback', function () {
      assert.strictEqual(true, reaperCb);
    });

    it('should be disabled', function () {
      assert.ok(T.hasOwnProperty('__ttl'));
      T.stopTTLReaper();
      assert.ok(!T.hasOwnProperty('__ttl'));

      T.create({ name: 'reaper killed' }, function (err, t) {
        assert.strictEqual(null, err);

        setTimeout(function () {
          T.collection.count(function (err, count) {
            assert.strictEqual(err, null);
            assert.equal(1, count);
            done();
          });
        }, interval + 10);

      });
    });
  });

  after(function () {
    describe('ttl index', function () {
      it('should be created', function (done) {
        assert.equal(true, indexCreated);
        T.collection.getIndexes({ full: true }, function (err, indexes) {
          if (err) return done(err);

          var index = indexes.filter(function (idx) {
            return '__ttl_1' === idx.name;
          })[0];

          assert.ok(index);
          assert.equal(true, index.background);
          mongoose.disconnect(done);
        });
      });
    });
  });

});


lib/ttl.js

/**
 * Module dependencies.
 */

var Model = require('mongoose').Model
  , ms = require('ms')

/**
 * Exports.
 */

module.exports = exports = ttl;

/**
 * Mongoose-TTL Plugin
 *
 * Provides timespan support for documents.
 *
 * Options:
 *   ttl: the time each doc should live in the db (default 60 seconds)
 *   interval: how often the expired doc reaper runs (default 5 mins)
 *   reap: enable the expired doc reaper (default true)
 *   onReap: callback passed to reaper execution
 *
 * Example:
 *
 *    var schema = new Schema({..});
 *    schema.plugin(ttl, { ttl: 5000 });
 *
 *  The ttl option supports the ms module by @guille meaning
 *  we can specify ttls with friendlier syntax. Example:
 *
 *    value     milliseconds
 *   ========================
 *    '2d'      172800000
 *    '1.5h'    5400000
 *    '1h'      3600000
 *    '1m'      60000
 *    '5s'      5000
 *    '500ms'   500
 *    100       100
 *
 * The expired document reaper can be disabled by passing `reap: false`.
 * Useful when working in multi-core environments when we only want one
 * process executing it.
 *
 *    var schema = new Schema({..});
 *    schema.plugin(ttl, { ttl: 5000, reap: false });
 *    var Cache = db.model('Cache', schema);
 *    if (isMyWorker) Cache.startTTLReaper();
 *
 *  The reaper can also be stopped.
 *
 *    Cache.stopTTLReaper();
 *
 *  Time-to-live is specified at the collection level, however
 *  it can also be overridden for a given document.
 *
 *    var cache = new Cache;
 *    cache.ttl = '2m' // lives for two minutes
 *    cache.save();
 *
 *  We can also reset the ttl for a given document to its
 *  default plugin state.
 *
 *    cache.resetTTL();
 *
 * @param {Schema} schema
 * @param {Object} options
 */

function ttl (schema, options) {
  options || (options = {});

  var key = '__ttl'
    , overridden = '__ttlOverride'
    , ttl = options.ttl || 60000
    , interval = options.interval || 60000*5
    , reap = false !== options.reap
    , onReap = 'function' == typeof options.onReap
      ? options.onReap
      : undefined

  var o = {};
  o[key] = Date;
  schema.add(o);

  schema.index(key, { background: true });

  schema.pre('save', function (next) {
    if (overridden in this) {
      // nothing to do
    } else {
      this[key] = fromNow();
    }
    next();
  });

  /**
   * startTTLReaper
   *
   * Starts reaping expired docs from the db.
   */

  schema.statics.startTTLReaper = function startTTLReaper () {
    if (key in this) return;

    var self = this;
    self[key] = setInterval((function remove () {
      self
      .remove()
      .where(key).lte(new Date)
      .exec(onReap);
      return remove;
    })(), interval);
  }

  /**
   * stopTTLReaper
   *
   * Stops removing expired docs from the db.
   */

  schema.statics.stopTTLReaper = function stopTTLReapter () {
    clearInterval(this[key]);
    delete this[key];
  };

  /**
   * Listen to Model.init.
   */

  schema.on('init', init);

  /**
   * init
   *
   * Hook into all model queries to include the TTL
   * filter and kick off the expired doc reaper if
   * enabled.
   * @private
   */

  function init (model) {
    if (model.__ttl) return;

    var distinct_ = model.distinct;
    model.distinct = function distinct (field, cond, cb) {
      applyTTL(cond);
      return distinct_.call(model, field, cond, cb);
    }

    'findOne find count'.split(' ').forEach(function (method) {
      var fn = model[method];

      model[method] = function (cond, fields, opts, cb) {
        if (!cond) {
          cond = {};
        } else if ('function' == typeof cond) {
          cb = cond;
          cond = {};
        }

        applyTTL(cond);
        return fn.call(model, cond, fields, opts, cb);
      }
    });

    'where $where'.split(' ').forEach(function (method) {
      var fn = model[method];
      model[method] = function () {
        var query = fn.apply(this, arguments)
          , cond = {};
        applyTTL(cond);
        return query.find(cond);
      }
    });

    if (reap) {
      model.startTTLReaper();
    }
  }

  /**
   * Getters/setters
   */

  var virt = schema.virtual('ttl');

  virt.get(function () {
    if (this[key]) return this[key];
    this.ttl = ttl;
    return this.ttl;
  });

  virt.set(function (val) {
    if ('reset' == val) return this.resetTTL();
    this[overridden] = arguments.length ? val : ttl;
    return this[key] = fromNow(this[overridden]);
  });

  /**
   * resetTTL
   *
   * Resets this documents ttl to the default specified
   * in the plugin options or plugin default.
   */

  schema.methods.resetTTL = function resetTTL () {
    delete this._doc[key];
    delete this[overridden];
  }

  /**
   * fromNow
   * @private
   */

  function fromNow (val) {
    var v = arguments.length ? val : ttl;
    return new Date(Date.now() + ms(v));
  }

  /**
   * Applies ttl to query conditions.
   * @private
   */

  function applyTTL (cond) {
    if (cond[key]) {
      cond.$and || (cond.$and = []);
      var a = {};
      a[key] = cond[key];
      cond.$and.push(a);
      var b = {};
      b[key] = { $gt: new Date };
      cond.$and.push(b);
      delete cond[key];
    } else {
      cond[key] = { $gt: new Date };
    }
  }
}

/**
 * Expose version.
 */

exports.version = JSON.parse(
    require('fs').readFileSync(__dirname + '/../package.json')
).version;
